File: dpp-backend/digitalproductpass/src/main/java/utils/EdcUtil.java
/*********************************************************************************
 *
 * Tractus-X - Digital Product Passport Application
 *
 * Copyright (c) 2022, 2024 BMW AG, Henkel AG & Co. KGaA
 * Copyright (c) 2023, 2024 CGI Deutschland B.V. & Co. KG
 * Copyright (c) 2022, 2024 Contributors to the Eclipse Foundation
 *
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the
 * License for the specific language govern in permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ********************************************************************************/

package utils;

import com.fasterxml.jackson.core.type.TypeReference;
import org.eclipse.tractusx.digitalproductpass.config.PolicyCheckConfig;
import org.eclipse.tractusx.digitalproductpass.config.PolicyCheckConfig.ActionConfig;
import org.eclipse.tractusx.digitalproductpass.config.PolicyCheckConfig.PolicyConfig;
import org.eclipse.tractusx.digitalproductpass.exceptions.ManagerException;
import org.eclipse.tractusx.digitalproductpass.models.edc.EndpointDataReference;
import org.eclipse.tractusx.digitalproductpass.models.general.Selection;
import org.eclipse.tractusx.digitalproductpass.models.negotiation.catalog.Dataset;
import org.eclipse.tractusx.digitalproductpass.models.negotiation.policy.Set;
import org.eclipse.tractusx.digitalproductpass.models.negotiation.policy.Constraint;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import utils.exceptions.UtilException;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * This class consists exclusively of methods to operate on Eclipse Dataspace Connector (EDC) data.
 *
 * <p> The methods defined here are intended to check or manipulate EDC's related data.
 */
@Component
public class EdcUtil {

    @Autowired
    private JsonUtil jsonUtil;

    @Autowired
    private PolicyUtil policyUtil;

    public EdcUtil() {
    }
    public EdcUtil(JsonUtil jsonUtil,PolicyUtil policyUtil) {
        this.jsonUtil = jsonUtil;
        this.policyUtil = policyUtil;
    }
    /**
     * Parses the data from HTTP request {@code Object} body to a {@code DataPLaneEndpoint} object data.
     * <p>
     *
     * @param body the {@code Object} object representing the body for the request.
     * @return a {@code DataPlaneEndpoint} object with the parsed date retrieved from the body.
     * @throws UtilException if unable to parse to the data.
     */
    public EndpointDataReference parseDataPlaneEndpoint(Object body) {
        try {
            return this.jsonUtil.bind(body, new TypeReference<>() {});
        } catch (Exception e) {
            throw new UtilException(EdcUtil.class, e, "It was not possible to parse the data plain endpoint");
        }
    }

    /**
     * Checks if the EDC is version v0.5.0 basing on the contractId format.
     * <p>
     *
     * @param contractId the contractId of the Digital Twin Registry (DTR) as a String.
     * @return true if the EDC's version is v0.5.0, false otherwise.
     * @throws UtilException if unable to check the EDC's version.
     */
    @SuppressWarnings("Unused")
    public Boolean isEdc5(String contractId) {
        try {
            String[] parts = contractId.split(String.format("\\%s", ":"));
            if (parts.length != 3) {
                throw new UtilException(EdcUtil.class, "It was not possible to check if EDC is v0.5.0, Invalid Contract Id");
            }
            return CrypUtil.isBase64(parts[1]); // If the contractId is base64 encoded
        } catch (Exception e) {
            throw new UtilException(EdcUtil.class, e, "It was not possible check if is the EDC v0.5.0");
        }
    }

    /**
     * Maps lists of datasets from a catalog into a map of contract ids and contracts
     * <p>
     *
     * @param datasets {@code List<Dataset>} the list of contracts to map.
     * @return {@code Map<String, Dataset>} the map of contract ids with the contracts
     * @throws UtilException if error when parsing the contracts
     */
    public Map<String, Dataset> mapDatasetsById(List<Dataset> datasets){
        try {
            return datasets.stream()
                    .collect(Collectors.toMap(Dataset::getId, Function.identity()));
        } catch (Exception e) {
            throw new UtilException(EdcUtil.class, e, "It was not possible to map the datasets");
        }
    }

    /**
     * Gets a valid contract from a map of contracts
     * <p>
     *
     * @param contracts {@code Map<String, Dataset>} the map of contracts by id
     * @param policyConfig {@code PolicyCheckConfig} the policy configuration to be checked
     * @return {@code Dataset} any valid contract from the map or null if no valid contract is valid
     * @throws UtilException if error when checking if contract is valid
     */
    public Dataset getValidContract(Map<String, Dataset> contracts, PolicyCheckConfig policyConfig){
        try {
            // Check all the contract policies against the configuration
            Map<String, Dataset> validContracts = this.filterValidContracts(contracts, policyConfig);
            if(validContracts == null || validContracts.size() == 0){
                return null; // No valid contract was found
            }
            String anyContractKey = validContracts.keySet().stream().findAny().orElse(null);
            return validContracts.get(anyContractKey); // Get any contract from the map
        } catch (Exception e) {
            throw new UtilException(EdcUtil.class, e, "It was not possible to check if contract is valid");
        }
    }
    /**
     * Gets a valid contract from a map of contracts
     * <p>
     *
     * @param contracts {@code Map<String, Dataset>} the map of contracts by id
     * @param policyConfig {@code PolicyCheckConfig} the policy configuration to be checked
     * @return {@code Dataset} any valid contract from the map or null if no valid contract is valid
     * @throws UtilException if error when checking if contract is valid
     */
    public Selection<Dataset, Set> selectValidContractAndPolicy(Map<String, Dataset> contracts, PolicyCheckConfig policyConfig){
        try {
            List<Selection<Dataset,Set>> validContractSelections  = new ArrayList<>(); // Array with possible selections
            contracts.keySet().forEach(key -> {
                Dataset contract = contracts.get(key); // Get the contract
                List<Set> policies = policyUtil.getValidPoliciesByConstraints(contract.getPolicy(), policyConfig); // Get all its valid policies
                // If policies are not null or are more than one
                if(policies == null || policies.size() == 0){
                    return;
                }
                policies.forEach(policy -> validContractSelections.add(new Selection<>(contract, policy))); // Create the possible selections
            });
            // Select any possible selection
            return validContractSelections.stream().findAny().orElse(null);
        } catch (Exception e) {
            throw new UtilException(EdcUtil.class, e, "It was not possible to get any valid contract and policy!");
        }
    }
    /**
     * Filters a map of contracts by policy configuration constraints
     * <p>
     *
     * @param contracts {@code Map<String, Dataset>} the map of contracts by id
     * @param policyConfig {@code PolicyCheckConfig} the policy configuration to filter the valid contracts
     * @return {@code Map<String, Dataset>} the map of contract ids with the valid contracts
     * @throws UtilException if error when parsing the contracts or filtering
     */
    public Map<String, Dataset> filterValidContracts(Map<String, Dataset> contracts, PolicyCheckConfig policyConfig){
        try {
            // If no configuration was provided all contracts are valid
            if(policyConfig == null){
                return contracts;
            }
            // If policy config is disabled then all the contracts are valid
            if(!policyConfig.getEnabled()){
                return contracts;
            }
            // Filter contracts by policy config
            return contracts.entrySet().stream()
                    .filter(contract -> this.isContractValid(contract.getValue(), policyConfig))
                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
        } catch (Exception e) {
            throw new UtilException(EdcUtil.class, e, "It was not possible to filter the valid contracts");
        }
    }

    /**
     * Checks if a contract is valid against the policy configuration
     * <p>
     *
     * @param contract {@code Dataset} the contract to be checked if valid
     * @param policyConfig {@code PolicyCheckConfig} the policy configuration to be checked
     * @return {@code Boolean} true if contract is valid
     * @throws UtilException if error when checking if contract is valid
     */
    public Boolean isContractValid(Dataset contract, PolicyCheckConfig policyConfig){
        try {
            return policyUtil.getValidPoliciesByConstraints(contract.getPolicy(), policyConfig).size() > 0; // If any policy is found the contract is valid
        } catch (Exception e) {
            throw new UtilException(EdcUtil.class, e, "It was not possible to check if contract is valid");
        }
    }
    /**
     * Get the contract valid policies as a list
     * <p>
     *
     * @param contract {@code Dataset} the contract to be checked if valid
     * @param policyConfig {@code PolicyCheckConfig} the policy configuration to be checked
     * @return {@code List<Set>} list of valid policies or empty list if no is available
     * @throws UtilException if error when checking if contract is valid
     */
    public List<Set> getContractValidPolicies(Dataset contract, PolicyCheckConfig policyConfig){
        try {
            // If no configuration was provided any contract is valid
            if(policyConfig == null){
                return new ArrayList<>();
            }
            // If policy config is disabled then the contract is valid
            if(!policyConfig.getEnabled()){
                return new ArrayList<>();
            }
            Object policies = contract.getPolicy();
            if (policies == null){
                throw new UtilException(EdcUtil.class,"The contract has no parsable policies");
            }
            // Check all the contract policies against the configuration
            return policyUtil.getValidPoliciesByConstraints(policies, policyConfig);
        } catch (Exception e) {
            throw new UtilException(EdcUtil.class, e, "It was not possible to check if contract is valid");
        }
    }

}

File: dpp-frontend/src/utils/edcUtil.js
/**
 * Tractus-X - Digital Product Passport Application
 *
 * Copyright (c) 2022, 2024 BMW AG, Henkel AG & Co. KGaA
 * Copyright (c) 2023, 2024 CGI Deutschland B.V. & Co. KG
 * Copyright (c) 2022, 2024 Contributors to the Eclipse Foundation
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the
 * License for the specific language govern in permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import jsonUtil from "@/utils/jsonUtil.js";

export default {
    parsePolicyConstraints(policy) {
        /**
         * Parses and simplifies the permissions, prohibitions and obligations of a given policy according to https://www.w3.org/ns/odrl/2
         * @param {object} policy - Contains the policy element that needs to be parsed and simplified
         * 
         * @returns {object} - Contains the permissions, prohibions and obligations of a policy normalized and parsed
         * @throws {Error} - When an excpetion is produced when parsint the policy
         */
        try {
            if (policy == null || policy == {} || policy.length == 0) {
                throw new Error("The incomming policy is empty!");
            }
            // Initialize policy constraints based on the odrl parameters
            // https://www.w3.org/ns/odrl/2/ODRL22.json
            let permissions = [];
            let prohibitions = [];
            let obligations = [];

            // Get policy constraints and normalize them based on the orld attributes
            permissions = this.parseProperties(jsonUtil.get("odrl:permission", policy, ".", []));
            prohibitions = this.parseProperties(jsonUtil.get("odrl:prohibitions", policy, ".", []));
            obligations = this.parseProperties(jsonUtil.get("odrl:obligations", policy, ".", []));

            // Returned normalized policy constraints for visualization
            return {
                "permissions": permissions,
                "prohibitions": prohibitions,
                "obligations": obligations
            }
        } catch (e) {
            throw new Error("It was not possible to parse the policy constraints [" + e.message + "]!");
        }
    },
    parseProperties(property, defaultValue = []) {
        /**
         * Parses and simplifies the permissions, prohibitions and obligations according to https://www.w3.org/ns/odrl/2
         * @param {object} property - Contains the permissions, prohibitions or obligations
         * @param {array|object|string} defaultValue - Default value to be returned when failed
         * 
         * @returns {object} - Contains the parsed constrains and the action type of the property
         */
        try {
            // Check for empty values and return default value
            if (property == null || property == {} || property.length == 0) {
                return defaultValue;
            }
            // Init variables
            let parsedProperty = []

            // If just one property is available do the parsing for just this value
            if (typeof property === 'object' && !Array.isArray(property)) {
                parsedProperty.push({
                    "actionType": this.parseActionType(jsonUtil.get("odrl:action", property, ".", null)),
                    "constraints": this.parseContraints(jsonUtil.get("odrl:constraint", property, ".", []))
                })
                return parsedProperty;
            }

            // Do the parsing for every value if more then one are available
            property.forEach((propertyValue) => {
                parsedProperty.push({
                    "actionType": this.parseActionType(jsonUtil.get("odrl:action", propertyValue, ".", null)),
                    "constraints": this.parseContraints(jsonUtil.get("odrl:constraint", propertyValue, ".", []))
                });
            });

            // Parse action type and the constraints associated with the permissions
            return parsedProperty;
        } catch (e) {
            console.error("It was not possible to parse the policy permissions [" + e.message + "]!");
            return defaultValue;
        }

    },
    parseActionType(actionType, defaultValue = null) {
        /**
         * Parses and simplifies the action type according to https://www.w3.org/ns/odrl/2
         * @param {object|string} actionType - Contains the action type object or string
         * @param {array|object|string} defaultValue - Default value to be returned when failed
         * 
         * @returns {string} - Simplied action type as string
         */
        try {
            // Check for empty values and return default value
            if (actionType == null || actionType == {} || actionType.length == 0) {
                return defaultValue;
            }
            // If attribute is an object it needs to contain the type
            if (typeof actionType === 'object' && !Array.isArray(actionType)) {
                return jsonUtil.get("odrl:type", actionType, ".", null)
            }
            // If the action is a string or value it needs to be added like it is
            return actionType;
        } catch (e) {
            console.error("It was not possible to parse the policy action type [" + e.message + "]!");
            return defaultValue;
        }
    },
    parseContraints(constraints, defaultValue = []) {
        /**
         * Parses and simplifies a list of one or more constaints with operator according to the https://www.w3.org/ns/odrl/2
         * @param {object} constraints - Contraints to be parsed with operator
         * @param {array|object|string} defaultValue - Default value to be returned when failed
         * 
         * @returns {object} - Object containing simplified contraint operator and list of constaints
         */
        try {
            // Check for empty values and return default value
            if (constraints == null || constraints == {} || constraints.length == 0) {
                return defaultValue;
            }
            // Initialized the values
            let parsedConstraints = [];
            let constraintsList = [];
            let operator = null;
            let possibleOperators = [
                "odrl:or",
                "odrl:and",
                "odrl:xone",
                "odrl:andSequence"
            ]

            // Get the first operator that appears (shall always be one available)
            operator = possibleOperators.find((selectedOperator) => jsonUtil.exists(selectedOperator, constraints));
            // No special operator is available so constraints will be in raw available
            if (!operator) {
                // Return no operator and the constraints
                let parsedConstraint = this.parseConstraint(constraints); //Parse the constraint
                if (!parsedConstraint || parsedConstraint == []) return; // If is null or not available do not store
                parsedConstraints.push(parsedConstraint); // Add to parsed constraint if it worked
                return {
                    "operator": null,
                    "constraint": parsedConstraints
                }
            }

            //Get list of constraints from operator
            constraintsList = jsonUtil.get(operator, constraints, ".", []);
            constraintsList.forEach((constraint) => {
                let parsedConstraint = this.parseConstraint(constraint); //Parse the constraint
                if (!parsedConstraint || parsedConstraint == []) return; // If is null or not available do not store
                parsedConstraints.push(parsedConstraint); // Add to parsed constraint if it worked
            });

            // Return the operator
            return {
                "operator": operator,
                "constraint": parsedConstraints
            }

        } catch (e) {
            console.error("It was not possible to parse the policy constraints [" + e.message + "]!");
            return defaultValue;
        }
    },
    parseConstraint(constraint, defaultValue = null) {
        /**
         * Parses and simplifies a constaint according to the https://www.w3.org/ns/odrl/2
         * @param {object} constraint - Contraint to be parsed
         * @param {array|object|string} defaultValue - Default value to be returned when failed
         * 
         * @returns {object} - Object containing simplified leftOperand, operator, and rightOperand as strings
         */
        try {
            // Check for empty constraint
            if (constraint == null || constraint == {} || constraint.length == 0) {
                return defaultValue;
            }
            //Initialize the variables
            let leftOperand = "";
            let rightOperand = "";
            let operator = "";

            //Check if the properties exist
            if (!jsonUtil.exists("odrl:leftOperand", constraint) || !jsonUtil.exists("odrl:rightOperand", constraint) || !jsonUtil.exists("odrl:operator", constraint)) {
                throw new Error("The constraint structure is not valid, because of missing attributes"); // Throw an exception that will be caught and handled as an error
            }

            //Check for operand types
            let rawOperator = jsonUtil.get("odrl:operator", constraint, ".", null);
            operator = rawOperator;
            if (typeof rawOperator !== "string") {
                operator = jsonUtil.get("@id", operator, ".", null)
            }

            // Parse the left operand
            let rawLeftOperand = jsonUtil.get("odrl:leftOperand", constraint, ".", null);
            leftOperand = rawLeftOperand;
            // If is not a string convert the object or array to string
            if (typeof rawLeftOperand !== "string") {
                leftOperand = jsonUtil.toJsonString(rawLeftOperand)
            }

            // Parse the right operand
            let rawRightOperand = jsonUtil.get("odrl:rightOperand", constraint, ".", null);
            rightOperand = rawRightOperand;
            // If is not a string convert the object or array to string
            if (typeof rawRightOperand !== "string") {
                rightOperand = jsonUtil.toJsonString(rawRightOperand)
            }

            // Return parsed constraint
            return {
                "leftOperand": leftOperand,
                "operator": operator,
                "rightOperand": rightOperand
            }

        } catch (e) {
            console.error("It was not possible to parse the policy constraint [" + e.message + "]!");
            return defaultValue;
        }
    }
};
